File: src/main/java/org/tby/fourdk/core/command/bus/CommandLoggingMiddleware.java
package org.tby.fourdk.core.command.bus;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tby.fourdk.core.command.Command;
import org.tby.fourdk.core.event.Event;

import java.util.List;

public class CommandLoggingMiddleware implements CommandBus {

    private static final Logger LOGGER = LoggerFactory.getLogger(CommandLoggingMiddleware.class);

    private CommandBus nextMiddleware;

    public CommandLoggingMiddleware(CommandBus commandBus) {
        this.nextMiddleware = commandBus;
    }

    @Override
    public List<Event> dispatch(Command command) {
        LOGGER.info("Handling command [{}]", command);
        var begin =  System.currentTimeMillis();
        var type = command.getClass().getSimpleName();
        List<Event> events;
        try {
            events = this.nextMiddleware.dispatch(command);
        } catch (RuntimeException e) {
            LOGGER.info("An error occurred while handling command [{}], exception: {}", command, e.getClass().getSimpleName());
            LOGGER.debug("An error occurred while handling command [{}], exception: {}", command, e);
            throw e;
        }
        var end = System.currentTimeMillis();
        var timeToProcessCommand = end - begin;
        LOGGER.info("Command of type [{}] [{}] has been handled in [{}] ms and has produced the following events [{}] [{}]",
                type, command, timeToProcessCommand, events.size(), events);
        return events;
    }


}

File: rust/project/core-rust/src/dddk/command/bus_impl/command_logging_middleware.rs
use log::{error, info};
use crate::dddk::aliases::Events;
use crate::dddk::command::command::Command;
use crate::dddk::command::command_bus::CommandBus;

pub struct CommandLoggingMiddleware {
    command_bus: Box<dyn CommandBus>,
}

impl CommandLoggingMiddleware {
    pub fn new(command_bus: Box<dyn CommandBus>) -> CommandLoggingMiddleware {
        CommandLoggingMiddleware {
            command_bus
        }
    }
}

impl CommandBus for CommandLoggingMiddleware {
    fn dispatch<'b>(&self, command: &'b dyn Command) -> Events {
        info!("Dispatching a command [{}] [{:?}].",
            command.get_command_name(),
            command);
        let events = self.command_bus.dispatch(command);
        if events.is_err() {
            let error = events.err().unwrap();
            let error_message = error.to_string();
            error!(
                "An error has occurred when dispatching command [{}] [{:?}]: {}",
                command.get_command_name(),
                command,
                error_message
            );
            return Err(error);
        }
        let events = events.unwrap();
        let mut event_names = String::new();
        events.iter()
            .for_each(|event| {
                event_names.push_str(event.get_event_name().as_str());
                event_names.push_str(" ");
            });
        info!(
            "Command[{}] [{:?}] has been handled and has produced [{}] events [{}].",
            command.get_command_name(),
            command,
            events.len(),
            event_names
        );
        return Ok(events);
    }
}

